### ISSUE TITLE
Push docker image after every commit to main?

### ISSUE THREAD

**Issue by wolfgangwalther** [PostgREST/postgrest#2006]
I'd like to pick up some of the discussion that was spread across different issues and PRs (#1857, #1957).

With the move to GitHub Actions we already have all the binaries available for download after every commit. It would be very helpful to have the docker image available at docker hub, after every commit to the main branch.

The current process of pre-releases requires manual intervention and is potentially error-prone:
- #1957 only changed the version in the `postgrest.cabal` file, but did not push a tag. No pre-release was built, although the current source suggests there should be one.
- Pushing tags without the requirement to merge to main, allows pushing arbitrary PRs - even with code that will never be merged. While this sounds beneficial at first to be able to test things before merging, this will also lead to us having multiple different long-living branches in the postgrest repo. And it potentially breaks the promise that each release "includes everything from all the other releases before that release". Suddenly `8.0.0-20211103` could have more or less features than `8.0.0-20211102` - or even different ones, when the two branches are next to each other.

In https://github.com/PostgREST/postgrest/issues/1857#issuecomment-955803907 @monacoremo wrote:

> > the only thing we'd need to do is upload the docker image
>
> We would need to upload to a separate repository to make sure that we don't compromise our dockerhub credentials. Pushing a tag for a pre-release is a reasonable middle ground I think [...]

Not compromising our dockerhub credentials is absolutely required, I agree. But is this really the case when only pushing **when merging to main**? This should basically be the same situation as right now, when a tag is pushed, no? Every merge is triggered by somebody with write privileges on the repo, where it's safe to use the dockerhub creds. Am I missing something?

**Comment by steve-chavez** [comment]
We would still need to change the `postgrest.cabal` version for replying with the correct version on the CLI and `Server` header though.

Another option could be pushing a docker image whenever the `postgrest.cabal` version(the 4th component) changes. Then the decision would be up to the commit author: "update postgrest.cabal version if you want a pre-release".

**Comment by monacoremo** [comment]
>> And it potentially breaks the promise that each release "includes everything from all the other releases before that release".
>
> True. It was also a bit weird doing the pre-release with github telling the commit doesn't exist on any branch.

Yes, that looks a bit weird. It's easy to check that the parent commit is on the `main` branch, though. For pre-releases I don't think that it's too critical, WDYT?

> We would still need to change the `postgrest.cabal` version for replying with the correct version on the CLI and `Server` header though.

Yes, this is what makes this a bit tricky. If we want to report a sensible version, we need to either patch the `postgrest.cabal` manually or automatically. Automatically was our previous approach to nightlies, but that was also a bit messy - it broke caches and also created releases that 'do not exist' and could not be reproduced. The manual way ensures that all builds are reproducible.

Tagging pre-releases is a good compromise I think - you can either update the version in a PR when you anticipate that you want to do a pre-release, or you update the version in a 'dangling' commit, pushed as a tag, that can be traced to a parent commit on `main`.

> Another option could be pushing a docker image whenever the postgrest.cabal version(the 4th component) changes. Then the decision would be up to the commit author: "update postgrest.cabal version if you want a pre-release".

This would be pretty smooth! It might be a bit brittle though, as CI only runs on the latest commit when multiple ones are pushed and we would need to check all of them for changes. We would also need to add a warning to `postgrest.cabal`. 

---

To be pragmatic, I propose that we:

* For testing of development versions:
  * We push docker images on every push to `main` as @wolfgangwalther suggests, tagged as `postgrest:main-[commit hash]`. 
    * Dockerhub credentials will be safe
    * The executable in the container will report an old version if the `postgrest.cabal` is not updated - but I think this is a risk we can take as the container is tagged as a dev version
  * For executables, we rely on the artifacts on each github-actions run
* To make pre-releases that we can link to in issues or elsewhere, we use the pre-release setup that is already in place 

WDYT?

**Comment by wolfgangwalther** [comment]
> > And it potentially breaks the promise that each release "includes everything from all the other releases before that release".
> 
> True. It was also a bit weird doing the pre-release with github telling [the commit doesn't exist on any branch](https://github.com/PostgREST/postgrest/commit/a15f988c33b067c283b1e7fa36d9a7add0fa7439).

I think we should just agree to not do that and to always commit the change to postgrest.cabal to main and always tag on main, too.

This will give us a consistent timeline for all releases and pre-releases, which has kind of the same or similar value as having reproducible builds in the first place.

**Comment by steve-chavez** [comment]
Not sure where to put this but I made a couple of dumb errors :clown_face: when creating a new pre-release:

- Messed up the date in `postgrest.cabal` https://github.com/PostgREST/postgrest/commit/3ded567f6df0ad399fa075fb1285c0a35f18fb32
- I pushed a tag without the `v` prefix(`9.0.0.20220107`) and wondered why nothing happened on CI for a while(good thing that it ignored it)

Maybe we can improve this process by having a `postgrest-pre-release` command in nix that automatically modifies the `postgrest.cabal` version with the current date and also creates a commit and tags it?

Or perhaps we can make that command a [do-nothing script](https://blog.danslimmon.com/2019/07/15/do-nothing-scripting-the-key-to-gradual-automation/) that just lists the steps? This one would be easier but also helpful.
